lib/mongo/collection.rb

```diff
@@ -537,19 +537,26 @@ def inspect
     # @since 2.0.0
     def insert_one(document, opts = {})
       client.send(:with_session, opts) do |session|
-        write_concern = write_concern_with_session(session)
-        write_with_retry(session, write_concern) do |server, txn_num|
-          Operation::Insert.new(
-              :documents => [ document ],
-              :db_name => database.name,
-              :coll_name => name,
-              :write_concern => write_concern,
-              :bypass_document_validation => !!opts[:bypass_document_validation],
-              :options => opts,
-              :id_generator => client.options[:id_generator],
-              :session => session,
-              :txn_num => txn_num
-           ).execute(server, client: client)
+
+      write_concern = if opts[:write_concern]
+        WriteConcern.get(opts[:write_concern])
+      else
+        write_concern_with_session(session)
+      end
```

Comment:
https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#behavior-of-the-writeconcern-field says the server will error when write concern is given in a transaction on an operation, therefore  I think the proposed code here is suitable.